ipa-core/src/protocol/ipa_prf/malicious_security/prover.rs

```diff
@@ -31,215 +81,264 @@ where
+    /// This function allows to set masks to protect sensitive information.
+    /// It is used when the last proof of a batch is generated.
+    ///
+    /// ## Panics
+    /// Panics when `self` contains more than one polynomial tuple
+    /// or when masks cannot be set safely.
+    /// The latter is the case when there are too many points for a polynomial of degree `R minus 1`.
+    /// This function likely panics when used with small fields like `Fp31`
+    /// Use `FP61BitPrime` instead when using this function in test cases.
+    pub fn set_masks(&mut self, p_0: F, q_0: F) {
+        // assert that there are only two polynomials
+        debug_assert_eq!(self.uv.len(), 1usize);
+        // assert that last point of p,q is really `F::Zero`
+        // such that we can safely set the 0 points to be the last points
+        // p is the polynomial uv_store.first().0
+        // and q is the polynomial uv_store.first().1
+        // this is likely to fail for small sized fields,
+        // so dont use Fp31 (but rather FP61BitPrime) in tests for this function.
+        debug_assert_eq!(
+            (self.first().0[R::USIZE - 1], self.first().1[R::USIZE - 1]),
+            (F::ZERO, F::ZERO)
         );
+        // shift 0 positions to the back where no actual point is stored
+        // (F::ZERO are overwhelmingly likely to be just 0 fillers)
+        self.uv[0].0[R::USIZE - 1] = self.uv[0].0[0];
+        self.uv[0].1[R::USIZE - 1] = self.uv[0].1[0];
+        // set 0 positions of p, q polynomial to masks
+        self.uv[0].0[0] = p_0;
+        self.uv[0].1[0] = q_0;
+    }
 
-        let mut p = GenericArray::<F, λ>::generate(|_| F::ZERO);
-        let mut q = GenericArray::<F, λ>::generate(|_| F::ZERO);
-        let mut proof: GenericArray<F, TwoNMinusOne<λ>> = GenericArray::generate(|_| F::ZERO);
-        for _ in 0..s {
-            for i in 0..λ::USIZE {
-                let x = u.next().unwrap_or(F::ZERO);
-                let y = v.next().unwrap_or(F::ZERO);
-                p[i] = x;
-                q[i] = y;
-                proof[i] += x * y;
+    /// This function generates a set of `UVPolynomial<F,R>` that is used for the next recursion.
+    /// Parameter `R` determines the degree of the polynomial which is `R-1`.
+    /// It also determines the amount of points that represent each polynomial which is `R`.
+    ///
+    /// The function uses `Fiat-Shamir` to generate a challenge `r`
+    /// by hashing the two input proofs, `proof_left` and `proof_right`
+    /// It then evaluates each input polynomial at x coordinate `r`.
+    /// The computed points are used to generate the set of `UVPolynomial<F,M>`.
+    ///
+    /// Uses Andy's borrow trick to work with references and owned values :)
+    pub fn gen_challenge_and_recurse<J, B>(
+        proof_left: &ZeroKnowledgeProof<F, R>,
+        proof_right: &ZeroKnowledgeProof<F, R>,
+        mut uv: J,
```

Comment:
No, because the first proof is not build from an UVStore which is a wrapper around a vector of field elements. The first proof is generated from an iterator over field elements since collecting this iterator in a vector would lead to a gigantic vector. This is also why I use Andy's borrow trick